### DISCUSSION TITLE
Jest problems with Expo 50

### DISCUSSION THREAD

**Discussion by abpbackup** [expo/expo#26856] [Archived: Questions]
There are some dependencies that are not using javascript as Jest expects and all the recommendations point to configurations on babel.config.js which was removed in expo 50. Here is the error I'm getting

Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    SyntaxError: /Users/andresbotero/projects/abp-soluciones/orion-nube-mobile/node_modules/@react-native/js-polyfills/error-guard.js: Missing semicolon. (14:4)

      12 | let _inGuard = 0;
      13 |
    > 14 | type ErrorHandler = (error: mixed, isFatal: boolean) => void;
         |     ^
      15 | type Fn<Args, Return> = (...Args) => Return;
      16 |
      17 | /**

      at constructor (node_modules/@babel/parser/lib/index.js:356:19)
      at Parser.raise (node_modules/@babel/parser/lib/index.js:3223:19)
      at Parser.semicolon (node_modules/@babel/parser/lib/index.js:3580:10)
      at Parser.parseExpressionStatement (node_modules/@babel/parser/lib/index.js:13158:10)
      at Parser.parseStatementContent (node_modules/@babel/parser/lib/index.js:12741:19)
      at Parser.parseStatementLike (node_modules/@babel/parser/lib/index.js:12588:17)
      at Parser.parseModuleItem (node_modules/@babel/parser/lib/index.js:12565:17)
      at Parser.parseBlockOrModuleBlockBody (node_modules/@babel/parser/lib/index.js:13189:36)
      at Parser.parseBlockBody (node_modules/@babel/parser/lib/index.js:13182:10)
      at Parser.parseProgram (node_modules/@babel/parser/lib/index.js:12464:10)
      at Parser.parseTopLevel (node_modules/@babel/parser/lib/index.js:12454:25)
      at Parser.parse (node_modules/@babel/parser/lib/index.js:14376:10)
      at parse (node_modules/@babel/parser/lib/index.js:14417:38)
      at parser (node_modules/@babel/core/lib/parser/index.js:41:34)
          at parser.next (<anonymous>)
      at normalizeFile (node_modules/@babel/core/lib/transformation/normalize-file.js:64:37)
          at normalizeFile.next (<anonymous>)
      at run (node_modules/@babel/core/lib/transformation/index.js:21:50)
          at run.next (<anonymous>)
      at transform (node_modules/@babel/core/lib/transform.js:22:33)
          at transform.next (<anonymous>)
      at evaluateSync (node_modules/gensync/index.js:251:28)
      at sync (node_modules/gensync/index.js:89:14)
      at stopHiding - secret - don't use this - v1 (node_modules/@babel/core/lib/errors/rewrite-stack-trace.js:47:12)
      at transformSync (node_modules/@babel/core/lib/transform.js:42:76)
      at ScriptTransformer.transformSource (node_modules/@jest/transform/build/ScriptTransformer.js:545:31)
      at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:674:40)
      at ScriptTransformer.transform (node_modules/@jest/transform/build/ScriptTransformer.js:726:19)
      at Object.<anonymous> (node_modules/react-native/jest/setup.js:314:6)

 FAIL  components/__tests__/AppHeader-test.js
  ● Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

**Comment by abpbackup** [comment]
https://github.com/react-native-community/upgrade-support/issues/152#issuecomment-1921986831

**Comment by vitorcamachoo** [comment]
I am getting this after updating to expo@50. Is this related?
```
Property 'toHavePathname' does not exist on type 'JestMatchers<{ update: (component: ReactElement<any, string | JSXElementConstructor<any>>) => void; unmount: () => void; rerender: (component: ReactElement<any, string | JSXElementConstructor<...>>) => void; ... 89 more ...; findAllByText: FindAllByQuery<...>; }>'.ts(2339)
```

**Reply by kishannareshpal** [reply]
Possibly. I've had the same type error for `toHavePathname` when upgraded to expo@50.

To resolve this I've done the following:

1. Open `expect-matchers.d.ts` which is located in the root of your project (if you don't have this file, then create it). Then modify to include the following lines:

```diff
# expect-matchers.d.ts

+ /// <reference types="expo-router/types/jest" />
+ declare module 'expect/build/matchers';
```

2. Open your `tsconfig.json` file and modify to include the following new line:

```diff
# tsconfig.ts

{
  "extends": "expo/tsconfig.base",
  "compilerOptions": {
    "strict": true,
    "baseUrl": ".",
    "paths": {
      "@/*": ["src/*"]
    }
  },
  "include": [
    "**/*.ts",
    "**/*.tsx",
    ".expo/types/**/*.ts",
    "expo-env.d.ts",
+    "expect-matchers.d.ts"
  ]
}
```

3. In your Jest configurations (e.g. `jest.config.ts` file) ensure that you have added `expo-router` and `escape-string-regexp` to the `transformIgnorePatterns`.

```diff
# jest.config.ts

import type { Config } from 'jest';

const config: Config = {
  preset: 'jest-expo',
  transformIgnorePatterns: [
#                                           👇                   👇
+    'node_modules/(?!((jest-)?react-native|escape-string-regexp|expo-router|@react-native(-community)?)|expo(nent)?|@expo(nent)?/.*|@expo-google-fonts/.*|react-navigation|@react-navigation/.*|@unimodules/.*|unimodules|sentry-expo|native-base|react-native-svg|@sentry/.*)'
  ],
  ...
};

export default config;

```

4. If you're using VS Code you may need to run the `TypeScript: Restart TS Server` (by opening the Command pallete using <kbd>CMD</kbd> / <kbd>CTRL</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd> then searching and running the command named previously. Or simply restart VS Code.

5. That's it.

---

Update (14th May 2024):
- I've corrected the step numbers as pointed out by @GregPetropoulos in his comment below. Thank you!

**Reply by GregPetropoulos** [reply]
### My Problem statement :
 Property 'toHavePathname' does not exist on type 'JestMatchers<{ update: (component: ReactElement<any, string | JSXElementConstructor<any>>) => void; unmount: () => void; rerender: (component: ReactElement<any, string | JSXElementConstructor<...>>) => void; ... 89 more ...; findAllByText: FindAllByQuery<...>; }>'.ts(2339)

### My Solution:
I was able to solve the issue of `toHavePathname` completely missing from the library. I couldn't even find it in the TS declarations. I solved it with only needing steps 1 and 2  of your solution.

***sidenote*** : In your step 4 which I think you ment step 3.
***sidenote*** : I am running this as a stand alone app with a EAS dev build (no expo go wrapper).


### Here is my test, if that helps others
<img width="1212" alt="Screenshot 2024-05-03 at 5 28 10 PM" src="https://github.com/expo/expo/assets/68525155/0465451b-efdb-4480-ab16-df42f72e8e54">

**Reply by kwburnett** [reply]
With Expo 52, I also only had to do steps 1 & 2, but I had to modify my `expect-matchers.d.ts` (which I didn't previously exist in the project) with the following:
```diff
# expect-matchers.d.ts

+ /// <reference types="expo-router/testing-library" />
```

This is with `jest-expo@52.0.4`.

**Reply by remyoudemans** [reply]
With Expo 53, I found that I had to do steps 1 & 2 but the content of the `expect-matchers.d.ts` file had to be
```diff
# expext-matchers.d.ts
/// <reference types="expo-router/types/expect" />
```

**Comment by rkstar** [comment]
i'm getting a module mapping error that i can't seem to resolve:

```
FAIL  src/shared/utils/tests/format/format-currency.test.ts
  ● Test suite failed to run

    Configuration error:

    Could not locate module  mapped as:
    <rootDir>/src/$1.

    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/^(.*)$/": "<rootDir>/src/$1"
      },
      "resolver": undefined
    }

      at createNoMappedModuleFoundError (node_modules/jest-resolve/build/resolver.js:759:17)
```

i've added the above to my `jest.config.js` but i'm still getting the error.

`jest.config.js`
```
const path = require('path');

// Sync object
module.exports = {
  preset: 'jest-expo',
  setupFilesAfterEnv: ['@testing-library/jest-native/extend-expect'],
  haste: {
    defaultPlatform: 'android',
    platforms: ['android', 'ios'],
  },
  transform: {
    '^.+\\.(js|jsx|ts|tsx)?$': './node_modules/babel-jest',
  },
  moduleNameMapper: {
    '/^(.*)$/': '<rootDir>/src/$1',
    '\\.svg': '<rootDir>/tests/mocks/svg.ts',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  rootDir: '.',
  moduleDirectories: [path.join('<rootDir>', 'node_modules')],
  modulePaths: ['<rootDir>/src/'],
  setupFiles: [
    '<rootDir>/tests/setup/index.js',
    '<rootDir>/node_modules/react-native-gesture-handler/jestSetup.js',
  ],
  transformIgnorePatterns: [
    'node_modules/(?!((jest-)?react-native|escape-string-regexp|expo-router|@react-native(-community)?)|expo(nent)?|@expo(nent)?/.*|@expo-google-fonts/.*|react-navigation|@react-navigation/.*|@unimodules/.*|unimodules|sentry-expo|native-base|react-native-svg|@sentry/.*)',
  ],
  testPathIgnorePatterns: ['local_modules'],
  testRegex: '(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js|jsx)$',
  globals: {
    'ts-jest': {
      tsconfig: 'tsconfig.spec.json',
      babelConfig: true,
    },
    __DEV__: true,
  },
  verbose: false,
};
```

**Reply by rkstar** [reply]
i've found that this problem is happening even when all i have in my `jest.config.js` file is `preset: jest-expo`

```
// package.json

"jest-expo": "^50.0.4",
```

```
// yarn.lock

"jest-expo@npm:^50.0.4":
  version: 50.0.4
  resolution: "jest-expo@npm:50.0.4"
  dependencies:
    "@expo/config": "npm:~8.5.0"
    "@expo/json-file": "npm:^8.2.37"
    "@jest/create-cache-key-function": "npm:^29.2.1"
    babel-jest: "npm:^29.2.1"
    find-up: "npm:^5.0.0"
    jest-environment-jsdom: "npm:^29.2.1"
    jest-watch-select-projects: "npm:^2.0.0"
    jest-watch-typeahead: "npm:2.2.1"
    json5: "npm:^2.2.3"
    lodash: "npm:^4.17.19"
    react-test-renderer: "npm:18.2.0"
    stacktrace-js: "npm:^2.0.2"
  bin:
    jest: bin/jest.js
  checksum: 10/00300c4f3a825fb57423b2f1cd3e90742aaee713b69610c3232c97c27ec7fb9636dc105a65e177ec762ca3c49b0528ac4d64010376c3d5c341cf709b9753be2c
  languageName: node
  linkType: hard
```

**Reply by rkstar** [reply]
after downgrading to `jest-expo: ^49.0.0` my tests run fine.

**Reply by jensdev** [reply]
@rkstar Did you manage to figure it out. I'm having the same issue

**Reply by vicwomg** [reply]
Also having this exact issue, anyone solve it? `jest-expo: ^49.0.0` runs fine and I'm not sure how to debug this any further.

**Reply by vicwomg** [reply]
For me, I was able to fix this issue by making this modification to my `tsconfig.json` not entirely sure why, but it worked:

<img width="443" alt="Screenshot 2024-06-14 at 1 10 11 PM" src="https://github.com/expo/expo/assets/4107190/4392bd02-8dbf-483f-800f-7ba4993a55a3">

**Comment by vanstinator** [comment]
Seeing the same on expo 50 and 51

**Comment by pfcodes** [comment]
Downgrading to `v49.0.0` resolved it for me on Expo 50
